### PR TITLE
API: XML is now generated without templates

### PR DESCRIPTION
**Pull request by annando**
There is now a generic functionality to generate XML output for API calls. Additionally some old code was dropped (support for a deprecated Activity Streams output). Also there are some changes to the sql queries.

There is some work left at the queries. But I wanted to do the pull request now so that the changes are split in several pull requests.

### CODE REVIEW

File: include/api.php [hunk 1]
```diff
@@ -681,59 +682,112 @@ function api_item_get_user(&$a, $item) {
...
+		$childname = key($data);
+		$data2 = array_pop($data);
+		$key = key($data2);
+
+		$namespaces = array("" => "http://api.twitter.com",
+					"statusnet" => "http://status.net/schema/api/1/",
+					"friendica" => "http://friendi.ca/schema/api/1/",
+					"georss" => "http://www.georss.org/georss");
+
+		/// @todo Auto detection of needed namespaces
```
Comment (fabrixxm) [inline include/api.php]: I think it's not needed. Just always put in namespace declarations and make the parser happy.. The parser don't complain on unused declared ns, but will stop on used undeclared ones...

**Comment by fabrixxm** [comment]
Have you tried this with some clients? 
I want to try with Friendly on Jolla, that uses xml

**Comment by annando** [comment]
I don't know clients that use the XML, but I compared every API call with an older Friendica server version and with a GNU Social server.

File: include/api.php (hunk 1 above)
Comment (annando) [inline include/api.php]: There are XML that never will return something with different namespaces (like the IDs of friends). I find it a little bit ugly to return namespaces there. Of course this is only some beautification.

Do you think this is a show stopper?

Comment (fabrixxm) [inline include/api.php]: > Do you think this is a show stopper?

absolutely not

**Comment by fabrixxm** [comment]
Ok, just give me some time to test it with Friendly :-)

**Comment by annando** [comment]
I will be so friendly :-)

**Comment by annando** [comment]
BTW: There had been massive differences between XML and JSON (missing fields and/or fields that aren't part of the JSON) and there were many functions that never had worked with XML output because of missing templates.

**Comment by rabuzarus** [comment]
Short note from my side (like always): Please add the description of the new db row to the db docu

**Comment by annando** [comment]
There is no new database row, just a new index.

**Comment by rabuzarus** [comment]
ahh ok. sorry

**Comment by annando** [comment]
@fabrixxm could you give me a hint how to install a sailfish emulator and this app? Then I could do some testing (for future changes) on my own. I want to do some more changes to the api in the future.

**Comment by tobiasd** [comment]
you can find the SailfishOS SDK stuff [here](https://sailfishos.org/wiki/Application_Development) and the Friendly source is [at Fabios projects repository](https://kirgroup.com/projects/fabrixxm/harbour-friendly). It's a customized QtCreator.

You need the i486 kit an run it then in debug mode,. Open the `friendly.pro` file and on the button below the help choose "Debug" -> "deploy as RPM package". QtCreator should tell you that the VM is not running and the emulator and ask it it should start them for you.

**Comment by annando** [comment]
@fabrixxm Have you done your testings?

**Comment by fabrixxm** [comment]
No, sorry, I'm quite busy right now. I'll test it this sunday.

**Comment by fabrixxm** [comment]
I was trying this PR but I have some problems because I'm on latest PHP (7.0.9) which raises a strange error:

> *Warning*:  Parameter 1 to api_statuses_home_timeline() expected to be a reference, value given in *include/api.php* on line *277*

Which is:

```php
$r = call_user_func($info['func'], $a, $type);
```

Looks like it want `$a` passed by reference because the function in `$info['func']` is defined as `api_statuses_home_timeline(&$a, $type)`. 

I get this warning also in current develop, looks like a php regression.

I never get any warning so far here. The only way I found to make it work is using

```php
$r = call_user_func_array($info['func'], array(&$a, $type));
```

That's made the code work in develop.

If I change that line in this patch, I get a new series of warning:

> *Warning*:  It is not yet possible to assign complex types to properties in *include/xml.php* on line *55*

**Comment by annando** [comment]
So we just should drop passing the "a" variable to the functions. I could include this in the current pull request if you like.

**Comment by fabrixxm** [comment]
Update:
I've made a test.
Looks like old PHP simply not complains and pass `$a` as value instead of reference when using `call_user_func()`.
The correct way is using `call_user_func_array()` with `$a` passed as reference to the array

I tried this
https://gist.github.com/fabrixxm/02dab6eadbbc6a93871b7d9bfcbb79bb
on php 5.6 and 7.0.2
Note that the output value should increase after every call to `my_function()` but doesn't after `call_user_func()`, without warning in old php and with warning in last php.

I would simply change that line to use

```php
$r = call_user_func_array($info['func'], array(&$a, $type));
```

After this, there is the new warning in `include/xml.php` :-D

**Comment by fabrixxm** [comment]
Ok, the warning is raised when there is an activity of an item. The array is: 
```
{
  "id" : ...,
  [...]
  "friendica_activities": {
      "like": [
        {
          "name": [
            "Mario Rossi"
          ],
          "url": [
            "http://friendica.local/profile/mrossi"
          ]
        }
      ],
      "dislike": [],
      "attendyes": [],
      "attendno": [],
      "attendmaybe": []
    },
```

So, every activity is an array of [user=>", url=>""] for every user that like or disliked or whatever the item.

The code at `include/xml.php:53-57` assume that if the key is a number, the value is a scalar, which in this case is false, because the value is an array.

As this is a friendica-only extension to the api output, we can decide how the xml should be, but I'm not sure on how to make the code work.

Using templates I was thinking something like:
```xml
<friendica:activities>
  <like>
    <contact>
        <user>Mario Rossi</user>
        <url>http://friendica.local/profile/mrossi</url>
    </contact>
  </like>
 ...
</activities>
```

(Note that the in the current develop branch, the XML still returns only the count of every sinlge activity, while the json is updated)

**Comment by annando** [comment]
The warning because of $a should now have disappeared.

Concerning the activities we should do it like Twitter once did (when they hadn't deprecated the XML output like they did now): there you have the plural and singular versions of the elements. Means: "statuses" and "status" or "users" and "user". We could do the same with "our" fields.

File: include/xml.php [hunk 2]
```diff
@@ -44,6 +47,15 @@ public static function from_array($array, &$xml, $remove_header = false, $namesp
 		}
 
 		foreach($array as $key => $value) {
+			if (!isset($element) AND isset($xml))
+				$element = $xml;
+
+			if (is_integer($key)) {
+				if (isset($element))
+					$element[0] = $value;
```
Comment (tobiasd) [inline include/xml.php]: for this I get the error `PHP Warning: It is not yet possible to assign complex types to properties in /.../include/xml.php on line 55` with PHP 7.0.9.

Posting new entries works (but raises the error) fetching the home time line does not (using Friendly).

**Pull request merged**